bin/run.py

```diff
@@ -400,9 +400,10 @@ def _get_expected_verdicts(self):
                 verdicts = [subdir]
             else:
                 if len(verdicts) == 0:
-                    error(
-                        f'Submission {self.short_path} must have @EXPECTED_RESULTS@. Defaulting to ACCEPTED.'
-                    )
+                    pass # TODO (Thore): made this shut up!
+                    #error(
+                    #    f'Submission {self.short_path} must have @EXPECTED_RESULTS@. Defaulting to ACCEPTED.'
+                    #)
```

Comment:
This error is probably still necessary. DOMjudge does not support an `expectations.yaml`, and reads the `@EXPECTED_RESULTS@` tag to know which final verdicts are allowed. So if the submission is not in one of the four default directories, and it does not have such a tag, DOMjudge will default to `ACCEPTED`, which is not intended in most cases :sweat_smile: 

I think that the `@EXPECTED_RESULTS@` tag converted to `expectations.yaml` syntax would look something like:
```yaml
rejected/broken.py:  # contains `@EXPECTED_RESULTS@: TIME_LIMIT_EXCEEDED, WRONG_ANSWER`
  permitted: [AC, TLE, WA]  # only these verdicts may appear
  required: [TLE, WA]       # at least one of these verdicts must appear
```
It would be nice if the existence of an `@EXPECTED_RESULTS@` could be checked for consistency with the `expectations.yaml`. Perhaps we could even infer the tag from the YAML on export, but that may result in too much seemingly magic behaviour?

Reply:
All correct, and all (in my mind) part of the _tool_ more than of the expecations framework.

Note that the framework does allow file-based expectations; in principle a submission _itself_ can supply “its own” expectations, which then compose with expectations in `expectations.yaml` in well-defined ways.

Reply:
By the way, if 
```yaml
permitted: [AC, TLE, WA] 
required: [TLE, WA]  
```
is sufficiently popular, the cleanest way is to elevate it to a common abbreviation next to `accepted`, `does not terminate`, etc. In principle, there are 
16 different possible `permitted` subsets (two of which make now sense), each of which can have a number of `required` subsets. I was just taking the standard set and add a few that seemed useful to me (such as `not accepted`).

If somebody has a large repository of @EXPECTED_RESULTS@-tagged files, we could run statistics on them! Please so.

Reply:
Running the following command on my folder that contains all Git repos that I've worked with in the past years (FPC from 2018, BAPC+preliminaries from 2020, NWERC from 2022):
```sh
$ grep -r @EXPECTED_RESULTS@ 2> /dev/null | grep 20 | cut -d ':' -f 3 | sed 's/ //g' | sort | uniq -c
```
- The `grep 20` is to only select repositories that have a year number (2018–2023), because I want to exclude things like BAPCtools and domjudge that use this tag many times in their tests.
- The `sed` command removes the optional space after the comma that separates the verdicts.
- The verdicts may still have different orders (e.g. both `ACCEPTED,WRONG_ANSWER` and `WRONG_ANSWER,ACCEPTED` occur), but I've manually aggregated those in the results below.

```
      3 ACCEPTED,RUN_TIME_ERROR
      2 ACCEPTED,RUN_TIME_ERROR,TIME_LIMIT_EXCEEDED
      2 ACCEPTED,RUN_TIME_ERROR,WRONG_ANSWER
     47 ACCEPTED,TIME_LIMIT_EXCEEDED
     16 ACCEPTED,WRONG_ANSWER
     10 RUN_TIME_ERROR,TIME_LIMIT_EXCEEDED
      3 RUN_TIME_ERROR,TIME_LIMIT_EXCEEDED,WRONG_ANSWER
     22 RUN_TIME_ERROR,WRONG_ANSWER
     49 TIME_LIMIT_EXCEEDED,WRONG_ANSWER
```

---

After running this command, I realize that my earlier example only works if the tag only has non-AC verdicts. For example, a combined `AC,TLE` verdict may better be written without `required`:
```yaml
mixed/sometimes-too-slow.py:  # contains `@EXPECTED_RESULTS@: ACCEPTED, TIME_LIMIT_EXCEEDED`
  permitted: [AC, TLE]  # only these verdicts may appear
```
Adding the `required` field is probably not even strictly necessary in _any_ of these cases, but for the cases that do not include `AC`, it's probably helpful if the tool warns about a submission that we expect to give some non-`AC` verdict, but is in fact `AC`.

---

And as a final remark, I'm finding it difficult to switch my mental model from "the verdict that the _submission_ receives" to "the set of verdicts that the _test cases_ receive". The fact that DOMjudge (and `@EXPECTED_RESULTS@`) use the former and `expectations.yaml` uses the latter, probably doesn't help either. :joy:

Reply:
Thank you for this. Also for this formulation:

>  I'm finding it difficult to switch my mental model from "the verdict that the submission receives" to "the set of verdicts that the test cases receive". 

That is _exactly_ why I’m doing this; and I’m trying to be terminologically stringent about it. To repeat this in my own words,

1. The separation of these two concepts is crucial for even _defining_ the semantics of, _e.g._, `time_limit_exceeded`, in accordance with the specification.
2. Another separation is between the behaviour of the  _construction and validation tool_ (which is used by authors) and the _electronic judge_ (which is used by contest organisers, teachers, and solvers.) The former (say, BAPCtools or problemtools) _needs_ a way of saying “no testcase for the brute-force solution can ever be `WA`, but `TLE` and `RTE` are acceptable, and the final verdict shan’t be `AC`”, and construction tools should support this to ensure problem quality. On the other hand, the judging environment (Kattis, DomJudge, etc.) needs well-defined semantics for what the _final_ verdict is (or maybe subtask scores of IOI), and when it is OK to abort prematurely to reduce server load. _For instance_ a judge can report the lexicographically first (by testcase path) non-`AC` verdict.

By the way, I’m happy to add `final` to the expectations specification, and tried many ideas. Here’s how that could look:

```yaml
final: [`TLE`]
permitted: [`AC`, `TLE`, `WA`]
required: [`TLE`]
```

This is then either a requirement about the lexicographic ordering of testcases, ensuring that the `TLE` verdict is encountered before the `WA` verdict (what the spec calls `first_error`), or that the judge has an ordering of rejection-verdicts that makes `TLE` worse than `WA`, continues grading after encountering `WA`, and then reports the worst error. As you can see.

But I don’t really see a need for it, nor is the semantics clear to me. This is the kind of feedback I’m soliciting here. (In my current BAPCtools integration, this issues would need to be resolved to understand whether the final verdict should be red or green.) See, for instance, submissions `bad-tle-wa.py` in the large screenshot above, which _fails_ (because it gets a `WA`), yet gets a green `TLE` (which, after all, is a permitted testcase verdict). I have no opinion on what the semantics or visualisation should be.

Reply:
Thanks for the elaboration! I don't think that a `final` expectation is needed, indeed :slightly_smiling_face:

Reply:
I’ve updated the pydoc documentation of `expectations.yaml` to spell out the workflow that avoids `expectation.yaml` altogether and specifies expectations submission-by-submission by writing
```python
registry['wrong_answer/greedy.py'] = Expectations({'sample': 'accepted'})
```
This would be my suggestion for translating submissions that contain `@EXPECTED_RESULTS@`. But I simply do not understand the semantics of the latter, because I can’t wrap my head around
> Use this to annotate the possible outcomes that a submission can have, that will be accepted by the Judging verifier. 
in the DOMjudge appendix.

[closed PR @ problem_package_format](https://github.com/Kattis/problem-package-format/pull/4)

Reply:
For DOMjudge, the final verdict not only depends on `result_prio`, but also on whether lazy judging is enabled: https://www.domjudge.org/docs/manual/8.2/judging.html#lazy-judging-and-results-priority

> When [lazy judging is] enabled, judging of a submission will stop when a highest priority result has been found for any testcase. You can find these priorities under the `results_prio` setting. In the default configuration, when enabling this, judging will stop with said verdict when a testcase results in e.g. `run-error`, `timelimit` or `wrong-answer`. When a testcase is correct (lower priority), judging will continue to the next test case. In other words, to arrive at a verdict of correct, all testcases will have been evaluated, while any of the ‘error’ verdicts will immediately return this answer for the submission and the other testcases will never be tested [...].
>
> [...]
> 
> When not using lazy judging, all testcases will always be ran for each submission. The `results_prio` list will then determine which of the individual testcase results will be the overall submission result: the highest priority one. In case of a tie, the first occurring testcase result with highest priority is returned.

I think we can assume the default DOMjudge settings for BAPCtools. At least, we have never changed this setting in DOMjudge in the past years :slightly_smiling_face: Note that the lazy judging, especially when running multiple runners in parallel, may not necessarily return the result of the first non-accepted test case. In the cases where we use `@EXPECTED_REULTS@`, I think we usually don't care about the ordering, as long as the submission never receives any other verdict than a verdict from this list (i.e. we don't use `first_error` currently, but I also don't think we knew that it existed). Thus, at least for now, I don't think we have to translate `first_error` to expectations.

I would translate `@EXPECTED_RESULTS@: <verdict_list>` as follows (and it would be nice if BAPCtools does this automatically, when an `expectations.yaml` does not exist):
- If it contains `AC`:
  ```yaml
  permitted: <verdict_list>
  ```
- If it does not contain `AC`:
  ```yaml
  permitted: [AC, *<verdict_list>]  # some test cases may pass, so also include `AC`
  required: verdict_list  # at least one of the specified verdicts must appear for at least one of the test cases
  ```